### ISSUE TITLE
Missed key presses in terminal rendering when working remotely with high latency

### ISSUE THREAD

**Issue by pfaion** [microsoft/vscode#126209]
Does this issue occur when all extensions are disabled?: Yes

_Note: VSCode Remote Development does not seem to work when starting with extensions disabled.
Instead, I used Extension Bisect (which does not seem to disable the remote extension) and the result was that the issue still occurs with all (other) extensions disabled._

- VS Code Version: 1.57.0
- OS Version: Darwin x64 19.6.0

**Context**
I work fully remotely with VSCode Remote Development via SSH.
The servers of my company are on the other side of the globe, so I have some natural SSH lag when working.
Sometimes there are spikes where the latency gets higher, in these moments I sometimes get incorrectly rendered text in the integrated terminal in VSCode.

**Steps to Reproduce:**
1. Have a remote setup with high latency
1. Type quickly in the terminal

Expected Result:
The terminal should show the exact result of the key sequence I typed (probably after a delay).

Actual Result:
Sometimes the terminal is missing a character (key press did not render) or erroneously shows a deleted character (backspace did not render). Some examples I had today:
- I type `git commit` and the terminal shows `gitcommit`.
- I type `c` `f` `<space>` `<backspace>` `<backspace>` `d` `<space>`, so I expect `cd `, but got `cfd `.

**If I reload the window, the terminal will actually display the correct text, so the key presses were registered and sent to the shell correctly! Just the rendering in the integrated terminal is bugged!**

Further insights from my investigations:
1. Some days it does not happen at all, sometimes I get these every 5 minutes.
1. If I open e.g. iTerm on my Mac and ssh into the remote server, I do not have these issues.
1. Since I have a complex ZSH setup with many plugins, I checked that the problem also happens with a stock bash shell in the integrated terminal. It seems to happen less often, but it also happened.
1. After the initial bug, continuing to use the terminal will lead to all kinds of weirdnesses, e.g. I was able to "delete" my prompt lead character by hitting backspace afterwards.
1. I have this issue already since I started working remotely (8 months), but hoped that others have the same problem and it would get fixed. But here I am now writing this report.

**Comment by Tyriar** [comment]
You can likely fix this by setting `"terminal.integrated.localEchoLatencyThreshold": -1`, you will lose the low latency local echo though.

**Comment by pfaion** [comment]
I think setting this has removed all the errors in the terminal for me with the cost of some higher latency (which is fine).

But I feel like this should **definitely not happen** (especially with default settings). Not seeing correctly what I typed in the Terminal is an absolute no go. I mean these cases really (just) annoy me, but I'm actually scared about the cases where I DO make a typo but the terminal shows incorrectly the text that I wanted to type (but didn't). That could lead to all kinds of unnoticed problems!

If you really want a "low latency response" over a high latency connection, I guess the better way would be to display some estimated text initially (as it happens now), but update that once you got the real value from the server? I don't know.

As a side note: I feel like I start seeing similar things just by editing files in vscode over a remote connection. Do you know if there's a similar mechanism at work there (and how to disable it)?

**Comment by pfaion** [comment]
Here's another really bad example of this happening (this is from a different machine, where I forgot to set up the `localEchoLatencyThreshold` fix):

```
git rebase -i HEAD~3                          
fatal: invalid upstream 'HEADD~3'
```

While I typed `git rebase -i HEAD~3`, I made a typo and quickly corrected it.

**And in the terminal now it looked correct!!!**

But as you can see in the response, it actually had `HEADD` instead of `HEAD`, but I would not see this!

As you can see, this can produce extreme security risks, because you end up executing different code than what you see!
Please consider changing this behavior to not be the default, fix it with better prediction or update the output continuously!

**Comment by fedorg** [comment]
I run into this annoying bug every day. The only thing that seems to help is completely disabling local echo like this.
`"terminal.integrated.localEchoLatencyThreshold": -1,`

**Comment by SoftwareApe** [comment]
Please please make this the default until the feature is confirmed to work. Also maybe that suggestion about eventual consistency should be implemented that @pfaion suggested, i.e. render the estimated contents until you have the real contents. There should be only a single source of truth here.

**Comment by davidcallanan** [comment]
I agree, the default should be changed until some better "eventual consistency" option is implemented.

If eventual consistency is ever implemented, I propose that there should be some indicator icon somewhere that will show whenever the terminal becomes consistent. So, you can choose to wait until this indicator appears before hitting return to make sure an incorrect command is not sent.

**Comment by injust** [comment]
I'm seeing behavior at high latency that I suspect is related to this issue. To reproduce:
- Set `terminal.integrated.localEchoLatencyThreshold` to 0 (or a higher value if that will enable local echo for your connection)
- Set `terminal.integrated.localEchoStyle` to `inverted`
- Open a terminal and mash on the keyboard

Local echo will trigger, and if you type fast enough, the text will become permanently inverted, even though text is supposed to switch back to un-inverted once acknowledged by the server.

**Comment by SamZhangQingChuan** [comment]
same issue here

**Comment by youngdjn** [comment]
Same issue here
